### PR TITLE
fix(iot-dev): Fix issue where client can't be re-opened if disconnected and retry policy is exhausted

### PR DESCRIPTION
**Pull request by timtay-microsoft**
IothubTransport layer did not notify DeviceIO layer when connection was lost for good, so it always believed to be open until closeNow is called. As a result, it would reject further attempts to open it even when the connection status callback reported as disconnected

This PR modifies the DeviceIO and IotHubTransport layer such that when the connection status in the transport layer changes to CONNECTED or DISCONNECTED, it executes a callback to the DeviceIO layer which then modifies its own state to stay consistent.

It also fixes a very small bug in the TransportClient class where it never set its own state to disconnected after calling close.

This PR also adds some basic state locks to the DeviceIO layer such that an open call and a close call cannot be executed simultaneously.

This PR also adds some e2e tests around recycling a device client including:

- Verify that a client object can be re-opened after closing it manually
- Verify that a client object can be re-opened from a main thread after a disconnection event exhausts the retry policy
- Verify that a client object can be re-opened from a thread spawned in the connection status callback thread after a disconnection event exhausts the retry policy

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java [hunk 1]
```diff
@@ -110,23 +110,17 @@
 
         /* Codes_SRS_DEVICE_IO_21_001: [The constructor shall store the provided protocol and config information.] */
         deviceClientConfigs.add(config);
-        this.config = config;
-        this.protocol = this.config.getProtocol();
-
-        /* Codes_SRS_DEVICE_IO_21_037: [The constructor shall initialize the `sendPeriodInMilliseconds` with default value of 10 milliseconds.] */
-        this.sendPeriodInMilliseconds = sendPeriodInMilliseconds;
-        /* Codes_SRS_DEVICE_IO_21_038: [The constructor shall initialize the `receivePeriodInMilliseconds` with default value of each protocol.] */
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: Why is this removed? Isn't it a breaking change?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java [hunk 2]
```diff
@@ -10,7 +10,7 @@
 import lombok.extern.slf4j.Slf4j;
 
 import java.io.IOException;
-import java.util.LinkedList;
+import java.util.ArrayList;
 import java.util.List;
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: Why array list?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java [hunk 3]
```diff
@@ -145,26 +139,37 @@
...
+                catch (DeviceClientException e)
+                {
+                    throw new IOException("Could not open the connection", e);
+                }
+            }
+
+            /* Codes_SRS_DEVICE_IO_21_014: [The open shall schedule receive tasks to run every receivePeriodInMilliseconds milliseconds.] */
+            /* Codes_SRS_DEVICE_IO_21_016: [The open shall set the `state` as `CONNECTED`.] */
+            commonOpenSetup();
+        }
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: is it ok to call commonOpenSetup with openTransport == false>?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java [hunk 4]
```diff
@@ -429,4 +445,32 @@ public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeC
...
+    {
+        try
+        {
+            //only close the device io layer, not the transport layer. The transport layer is already closed when this method is called
+            this.close(false);
+        }
+        catch (IOException e)
+        {
+            log.warn("Encountered an exception while closing device IO layer due to transport disconnection event", e);
+        }
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: This is bad pattern to catch exception and only log. it is absorbing the exception. What action can SDK user perform in this situation?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/TransportConnectionListener.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ */
+
+package com.microsoft.azure.sdk.iot.device;
+
+public interface TransportConnectionListener
+{
+    public void onTransportConnectionLost();
+
+    public void onTransportConnectionRecovered();
+}
```
Comment (prmathur-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/TransportConnectionListener.java]: 👍

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java (hunk 1 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: This was duplicate code within this method. These values were set to the same value twice

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java (hunk 2 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: David recommended this change. Arraylist performs better at get/set operations, and linkedlist performs better at add/remove operations. Since we are getting more often then adding, arraylist is the better choice

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java (hunk 3 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: commonOpenSetup doesn't do anything to the transport layer, so it wouldn't be useful as a parameter regardless

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java (hunk 4 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: This exception is only caught when opening/closing the transport layer. Since the transport layer isn't being opened/closed, this exception should never be caught

Comment (prmathur-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: In that case would it be safe to rethrow after logging?

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java [hunk 6]
```diff
@@ -281,6 +281,68 @@ public void sendMessagesWithTcpConnectionDropNotifiesUserIfRetryExpires() throws
...
+    {
+        if (testInstance.protocol == HTTPS || testInstance.useHttpProxy)
+        {
+            return;
+        }
+
+        this.testInstance.setup();
+
+        this.testInstance.client.setRetryPolicy(new NoRetry());
+
```
Comment (prmathur-microsoft) [inline iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: Can you also add a test for retrying and recovering by explicitly setting lower retry timeout?

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java [hunk 7]
```diff
@@ -281,6 +281,68 @@ public void sendMessagesWithTcpConnectionDropNotifiesUserIfRetryExpires() throws
...
+    public void clientCanReopenAfterDisconnectionEvent() throws Exception
+    {
+        if (testInstance.protocol == HTTPS || testInstance.useHttpProxy)
+        {
+            return;
+        }
+
+        this.testInstance.setup();
+
+        this.testInstance.client.setRetryPolicy(new NoRetry());
```
Comment (abhipsaMisra) [inline iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: We would also need tests where the device client has a set retry policy; and the application reconnects and recovers on retry policy expiry.

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java (hunk 6 above)
Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: So that the retry policy gets exhausted, and then application level retry is successful? Sure

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java (hunk 7 above)
Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: I'll add that, and a similar test for when the device operation timeout causes the disconnection

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java (hunk 6 above)
Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: Actually, this kind of test isn't possible to write right now. The fault injection framework we use does not allow us to cause a retryable error for X number of seconds. That functionality is necessary for us to test running through the sdk level retry and some application level retry. We only get DISCONNECTED_RETRYING followed by a successful reconnect within the SDK retry logic

File: iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java (hunk 7 above)
Comment (timtay-microsoft) [reply iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/tests/iothubservices/errorinjection/SendMessagesErrInjTests.java]: Actually, this kind of test isn't possible to write right now. The fault injection framework we use does not allow us to cause a retryable error for X number of seconds. That functionality is necessary for us to test running through the sdk level retry and some application level retry. We only get DISCONNECTED_RETRYING followed by a successful reconnect within the SDK retry logic

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java (hunk 4 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java]: Upon reworking the DeviceIO layer, this is no longer an issue. No exceptions are being swallowed like this anymore.

**Comment by timtay-microsoft** [comment]
Closing this PR as I had to redesign the DeviceIO layer to fix an issue in this PR where the DeviceIO layer leaked threads. The correct fix for this bug can be seen in #608
